Add support for saving and restoring sessions from a database.

lotusronin opened the pull request:
A session is defined be a username, an auth token, and a transaction id.
The CLI app now displays any saved sessions and allows them to be used
for login.

An SessionDb.sq file was added and is used to generate a database class and query functions.
Each platform provides an implementation of the DriverFactory class which the UIs will pass in to the shared code which is used to make an instance of the generated database class. This database class is held in a singleton object which is the interface the rest of the shared code will use.

Additionally, a transaction id is now passed to the MatrixSession constructor and it is used and incremented each message send.

Addresses #16.

lotusronin committed:
Add support for saving and restoring sessions from a database.

A session is defined be a username, an auth token, and a transaction id.
The CLI app now displays any saved sessions and allows them to be used
for login.

Limvot commented on serif_cli/src/main/kotlin/serif_cli/App.kt:
```diff
@@ -29,19 +30,52 @@ class App {
+        //Each UI will create it's specific DriverFactory
+        //And call this function before the backend can get
+        //information out of the database
+        Database.initDb(DriverFactory())
```
I think the shared code can create the driver itself, right? It feels a bit out of place in the UI.

lotusronin replied:
You would be mostly right, unfortunately the android driver requires a Context var to be passed in, so we would need to pass the android context to the backend.

Limvot replied:
gah that's a good point. Android man

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt:
```diff
@@ -106,9 +110,38 @@ class MatrixClient {
+        //Load from DB
+        println("loading specific session from db")
+        val sessions = Database.getStoredSessions()
+        for((user, tok, transactionId) in sessions) {
```
Can we get the right session from the database directly, instead of getting all of them and looping?

lotusronin replied:
I imagine so, I can fiddle with the SQL to try and add a query by name.

lotusronin replied:
Updated with a session select by user.

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt:
```diff
@@ -106,9 +110,38 @@ class MatrixClient {
+        return Error("No Saved Session for $username")
+    }
+
+    fun getStoredSessions() : List<String> {
```
Should this be a property instead of a getter? Might be part of a larger style discussion

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/States.kt:
```diff
@@ -17,6 +17,17 @@ class MatrixLogin(val login_message: String, val mclient: MatrixClient): MatrixS
+        }
+    }
+
+    fun getSessions() : List<String> {
```
Same property thought

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/States.kt:
```diff
@@ -48,5 +59,11 @@ class MatrixChatRoom(val msession: MatrixSession, val room_id: String, val messa
         return MatrixRooms(msession, msession.rooms, "Back to rooms. You can still do a :sync...")
     }
+
+    fun getRoomName(): String {
```
same property thought - also, maybe Room should get it's name on creation, so it doesn't have to re-lookup?

lotusronin replied:
Saving the room name is definitely the way to go.

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/db/Db.kt:
```diff
@@ -0,0 +1,40 @@
+}
+
+fun saveSession(db : SessionDb?, username : String, access_token : String, transactionId : Long) {
+    if(db != null) {
```
These could be made a bit more compact with the nullability operators

db?.sessionDbQueries?.insertSession(username, access_token, transactionId)

lotusronin replied:
That's slick.

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/db/Db.kt:
```diff
@@ -0,0 +1,40 @@
+    }
+}
+
+fun updateSession(db : SessionDb?, auth_tok : String, transactionId : Long) {
```
Same idea as saveSession, with the nullability operators

lotusronin replied:
That's still slick.

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/db/Db.kt:
```diff
@@ -0,0 +1,40 @@
+}
+
+fun getStoredSessions(db : SessionDb?) : List<Triple<String,String,Long>> {
+    if(db != null) {
```
Even more nullability operators!

lotusronin replied:
That's yet again slick.

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/Database.kt:
```diff
@@ -0,0 +1,29 @@
+import xyz.room409.serif.serif_shared.db.*
+import xyz.room409.serif.serif_shared.db.DriverFactory
+
+object Database {
```
This just seems to pass everything through, can we just have the implementation here?

lotusronin replied:
It does just pass through, really the key difference is the package the file is a part of. This Database file is part of the `xyz.room409.serif.serif_shared` package, while the other file is part of the `xyz.room409.serif.serif_shared.db` package, along with all of the generated code from sqldelight. My initial thought was have the methods for interfacing with the generated code all tucked away in the one file that was in the `.db` package (Db.kt), and then provide `shared_serif` with an interface to it (Database.kt), but we already have 2 pass throughs happening with the state and client functions and there wasn't any additional cleanup needing to be done between the generated code and the shared code so moving the implementation up one level makes sense.

Limvot reviewed (commented):
I dig the database stuff, but have a few questions about moving code around and some style things. Dig the direction overall!

lotusronin committed:
Code cleanup to address PR feedback.

Cleanup null check with the ?. operator, pulled in implementation
details for the database up to the Database object. Pass in room
name to the Room state to avoid repeated lookups.

lotusronin committed:
Add a query for a session by user

lotusronin committed:
No need for getRoomName, just do a property read
